### PR TITLE
[Model] Video Frame Sparsification via Pixel-Level Similarity for Efficient Multimodal Long-Video Inference for Qwen2.5_vl and Qwen3_vl

### PR DESCRIPTION
**Pull request by rayleeku**
## Purpose

RFC: https://github.com/vllm-project/vllm/issues/31803

The exponential growth of video content and the rising demand for long-video intelligence (e.g., video summarization, multimodal QA, video generation) have made long-video inference a standard requirement for multimodal large models. Current models suffer from critical efficiency bottlenecks and face two key challenges: temporal and spatial redundancy of videos and input sequence overflow, thus necessitating a lightweight, adaptive, and scalable sparsification solution. This PR proposes a video sparsification scheme that adaptively selects key frames based on video inter-frame similarity and user-defined sparsity rates, and it has been compatible with Qwen2.5-VL and Qwen3-VL.

## Implementation details
This solution implements adaptive sparsification through three stages:
1. Downsampling Preprocessing: Apply spatial downsampling (e.g., 4× reduction) to raw frames to lower computational load.
2. Similarity Assessment: Compute SSIM + L1 loss between adjacent frames to generate inter-frame similarity sequences.
3. Adaptive Keyframe Extraction:
    Calculate retained frame count: K = int(total_frames × sparsity_ratio) based on user-defined sparsity ratio
    Apply constraint: K = max(2, min(total_frames, 2 × (K // 2))) → ensures K is even and ≥2
    Select Top-(K-1) lowest-similarity frames as segmentation points to divide the video into K segments
    Extract the temporal midpoint frame from each segment as the keyframe; discard others.

## Test Plan
- [ ] Add the variable video_sparse_rate when launching the service.
- [ ] Test the performance and functionality of Qwen2.5-VL/Qwen3-VL separately when video input is provided， online and offline.
## Test Result
q3
q2

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/inputs/preprocess.py [hunk 1]
```diff
@@ -38,7 +38,7 @@
     token_inputs,
 )
 from .parse import is_explicit_encoder_decoder_prompt, parse_singleton_prompt
-
+from vllm.multimodal.video_sparse import SimilarFrameDetector, is_multimodal_EFS_enabled
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py]: Please rename `EFS` to `efs` to conform with snake case

File: vllm/engine/arg_utils.py [hunk 2]
```diff
@@ -1282,6 +1286,7 @@ def create_model_config(self) -> ModelConfig:
             logits_processors=self.logits_processors,
             video_pruning_rate=self.video_pruning_rate,
             io_processor_plugin=self.io_processor_plugin,
+            video_sparse_rate=self.video_sparse_rate,
```
Comment (DarkLight1337) [inline vllm/engine/arg_utils.py]: Can you group this with `video_pruning_rate`?

File: vllm/multimodal/video_sparse.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+        use_downsampled_loss: bool = True,
+        downscale_factor: int = 4,
+        alpha: float = 0.85,
+    ):
+        self.sparse_ratio = sparse_ratio
+        self.use_downsampled_loss = use_downsampled_loss
+        self.downscale_factor = downscale_factor
+        self.alpha = alpha
+
+    def _detect_and_convert_format(
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: You can use `transformers.image_utils.infer_channel_dimension_format` which works on videos as well

File: vllm/multimodal/video_sparse.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+        Use torch interpolate (GPU) or cv2 resize (CPU fallback).
+        """
+        frame_number, channels, height, width = frames.shape
+        new_height, new_width = (
+            height // self.downscale_factor,
+            width // self.downscale_factor,
+        )
+
+        # Use torch interpolate if available (GPU-optimized)
+        if hasattr(torch.nn.functional, "interpolate"):
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: In what case does `torch.nn.functional.interpolate` not exist?

Comment (DarkLight1337) [reply vllm/multimodal/video_sparse.py]: I suggest using the built in methods of `BaseVideoProcessor` from Transformers library to perform resizing

File: vllm/multimodal/video_sparse.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        # Move to CPU for cv2 operations
+        gray1_np = gray1.cpu().numpy() if gray1.device.type != "cpu" else gray1.numpy()
+        gray2_np = gray2.cpu().numpy() if gray2.device.type != "cpu" else gray2.numpy()
+
+        # SSIM constants
+        C1 = (0.01 * 255) ** 2
+        C2 = (0.03 * 255) ** 2
+
+        # Gaussian blur for mean/variance calculation
+        mu1 = cv2.GaussianBlur(gray1_np, (11, 11), 1.5)
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: `torchvision` also supports gaussian blur so I think there is no real need to use opencv unless it's faster for CPU input

File: vllm/engine/arg_utils.py (hunk 2 above)
Comment (rayleeku) [reply vllm/engine/arg_utils.py]: I will group them together. Thanks

File: vllm/multimodal/video_sparse.py (hunk 3 above)
Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: Thanks for your comments. I compared two functions. The `transformers.image_utils.infer_channel_dimension_format` function can only return the image format and does not return the transformed data. Therefore, I tend to prefer using a custom new function.

File: vllm/multimodal/video_sparse.py (hunk 5 above)
Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: I have replaced the code with a torchvision implementation. Thanks

File: vllm/multimodal/video_sparse.py (hunk 4 above)
Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: I used the function `torchvision.transforms.v2.functional.resize`, which is identical to the underlying resize function in `BaseVideoProcessor`, to perform the resizing. This approach eliminates the need to initialize a `BaseVideoProcessor` instance.
